mirascope/chat/parsers.py

```diff
@@ -0,0 +1,66 @@
+"""Classes for using parsers with Chat APIs."""
+import json
+from typing import Callable, Generator, Optional, Type, Union
+
+from openai.types.chat.chat_completion_message_tool_call import (
+    ChatCompletionMessageToolCall,
+    Function,
+)
+from pydantic import BaseModel
+
+from .tools import OpenAITool
+from .types import OpenAIChatCompletionChunk
+from .utils import convert_tools_list_to_openai_tools
+
+
+class OpenAIToolStreamParser(BaseModel):
+    """A utility class to parse `OpenAIChatCompletionChunk`s into `OpenAITools`."""
+
+    tool_calls: list[ChatCompletionMessageToolCall] = []
+    tools: list[Union[Callable, Type[OpenAITool]]] = []
+
+    def from_stream(
+        self, stream: Generator[OpenAIChatCompletionChunk, None, None]
+    ) -> Generator[OpenAITool, None, None]:
+        """Parses a stream of `OpenAIChatCompletionChunk`s into `OpenAITools`."""
+        openai_tools = convert_tools_list_to_openai_tools(self.tools)
+        tool_type: Optional[Type[OpenAITool]] = None
+        for chunk in stream:
+            if not chunk.tool_calls:  # is true for first and final chunk
+                continue
+            tool_call_chunk = chunk.tool_calls[
+                0
+            ]  # update if multiple tool calls are supported
+            if len(self.tool_calls) <= tool_call_chunk.index:
```

Comment:
break up the below into sections with header comments explaining the steps in plain text (not too verbose)